File: 187-repeated-dna-sequences/repeated-dna-sequences.cpp
class Solution {
public:
    int convert(char c){
        if(c=='A') return 0b00;
        if(c=='C') return 0b01;
        if(c=='T') return 0b10;
        return 0b11;
    }
    vector<string> findRepeatedDnaSequences(string s) {
        int encode=0;
        unordered_map<int,int> seen;
        vector<string> result;
        for(int i=0;i<s.size();++i){
            encode=((encode<<2) | convert(s[i]))&0xFFFFF;
            if(i>=9){
                if(++seen[encode]==2) result.push_back(s.substr(i-9,10));
            }
        }
        return result;
    }
};

File: 187-repeated-dna-sequences/repeated-dna-sequences.py
class Solution:
    def findRepeatedDnaSequences(self, s: str) -> List[str]:
        n=len(s)
        
        def binary(char):
            match char:
                case "A":
                    return 0b00
                case "C":
                    return 0b01
                case "G":
                    return 0b10
            
            return 0b11

        mask=0
        seen=defaultdict(int)
        sol=[]

        for i in range(n):
            mask=((mask<<2)|binary(s[i]))&0xFFFFF
            if i>=9:
                if seen[mask]==1:
                    sol.append(s[i-9:i+1])
                
                seen[mask]+=1
        
        return sol